### PR TITLE
Presupplied ORFs functionality

### LINKED ISSUE #255: Allow input of pre-annotated ORFs
### Description of feature

Originating from e.g. metatdenovo and mag pipelines. This would be very helpful for me as I run those pipelines and would like to keep the ORF ids for all analyses.

### PR DESCRIPTION
**Pull request by jfy133**
Closes #255 

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/funcscan/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/funcscan _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: bin/check_samplesheet.py [hunk 1]
```diff
@@ -85,13 +107,44 @@ def _validate_fasta(self, row):
...
+        """Assert that a given filename has one of the expected FASTA extensions."""
         filename = Path(row[self._contig_col]).name
         assert any(filename.endswith(extension) for extension in self.VALID_FORMATS), (
             f"The FASTA file has an unrecognized extension: {filename}\n"
             f"It should be one of: {', '.join(self.VALID_FORMATS)}"
         )
 
+    def _validate_protein(self, row):
+        """Assert that the amino acid FASTA entry has the right format."""
+        if len(row[self._protein_col]) > 0:
```
Comment (Midnighter) [inline bin/check_samplesheet.py:119]:
```suggestion
        if self._protein_col in row and len(row[self._protein_col]) > 0:
```

File: bin/check_samplesheet.py [hunk 2]
```diff
@@ -85,13 +107,44 @@ def _validate_fasta(self, row):
...
+        filename = Path(row[self._protein_col]).name
+        if len(row[self._protein_col]) > 0:
+            assert any(filename.endswith(extension) for extension in self.VALID_PROTEIN_FORMATS), (
+                f"The protein FASTA file has an unrecognized extension: {filename}\n"
+                f"It should be one of: {', '.join(self.VALID_PROTEIN_FORMATS)}"
+            )
+
+    def _validate_feature(self, row):
+        """Assert that the feature file entry has the right format."""
+        assert len(row[self._feature_col]) > 0 and (
```
Comment (Midnighter) [inline bin/check_samplesheet.py:135]:
```suggestion
        if self._feature_col in row:
            assert len(row[self._feature_col]) > 0 and (
```

**Review by Midnighter — COMMENTED** [review]
These two changes should make the columns optional. Maybe you can add a comment making that clear.

**Comment by jfy133** [comment]
Update: maybe make it possible to run _just_ annotation (to get rid of that horrible if/else statement)?

**Comment by jfy133** [comment]
Once https://github.com/nf-core/funcscan/pull/285 is merged, should sync with dev and replace the samplesheet input. This should make this much easier with input validation, which keeps breaking for some reason

**Comment by jfy133** [comment]
I'm going to start from scratch as there are a lot of improvements that make this better

**Pull request closed without merging**